### PR TITLE
Added CORBA blog.

### PR DESCRIPTION
**Pull request by mwhittaker**

### CODE REVIEW

File: website/blog/corba.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+  called as
+  [`f.remote`](https://docs.ray.io/en/latest/ray-core/walkthrough.html#running-a-task)
+  when invoked remotely.
+- However, placing a syntactic burden on developers to make remote method calls
+  is unnecessary. Even [the paper][a_note_on_dc] which describes why unifying
+  local and remote objects is doomed to fail does "not argue against pleasant
+  programming interfaces."
+- Some people argue that a syntactic marker for remote method calls is needed to
+  see that a remote call is happening. However, any syntactic marker is
+  inadequate, as a remote method call can be easily wrapped inside a function,
```
Comment (ghemawat) [inline website/blog/corba.md:59]: Maybe change "can be easily wrapped" to something like "is usually wrapped".

And I wonder if we can find some common piece of OSS code that we can point at as an example where a call looks local bit it wraps inside it something that is remote.

Also we could point out that DB calls are sometimes local and sometimes remote (sqlite backend for Go vs. say Postgres).

File: website/blog/corba.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+  and callers of this function now don't have the syntactic marker to see a
+  remote method call is happening under the covers.
+- Service Weaver remote method calls are syntactically identical to local method
+  calls. They do **look** the same, and this is okay. TODO: Give a short
+  example.
+- NOTE(mwhittaker): There might be an argument that using futures makes it
+  syntactically obvious a remote call could be happening, and futures cannot
+  easily be erased by wrapping them in functions. I think it might be too fine
+  of a point to bring up?
```
Comment (ghemawat) [inline website/blog/corba.md:68]: Futures are used for other things than remote calls:
- work done on another thread
- IO (even local)
- Sleeps / timers

And this in turn infects all the way up and pretty much everything turns into returning futures (or being marked async) and it again becomes hard to tell what is local and what is remote.

I don't think it is worth mentioning futures.

**Review by ghemawat — COMMENTED** [review]
Very nice writeup. I like how it succintly hits the relevant points. Thanks.

File: website/blog/corba.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+## Acting Local
+
+- When we say a remote method call **acts** like a local method call, we mean
+  that it exhibits the same **semantics**.
+- This is where frameworks get into trouble. Remote method calls fundamentally
+  have different semantics than local method calls. This is where the fallacies
+  of distributed computing come in to play.
+- For example, remote method calls execute with significantly higher latency
```
[the remark this answers is not in the document]
Comment (mwhittaker) [reply website/blog/corba.md]: Done.

File: website/blog/corba.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+  than local method calls. Remote method calls also exhibit partial (and
+  sometimes undetectable) failures. Local method calls, on the other hand, can
+  leverage fate sharing; if something crashes, everything crashes.
+- Because of these fundamental semantic differences, trying to make a remote
```
[the remark this answers is not in the document]
Comment (mwhittaker) [reply website/blog/corba.md:78]: Added! [A Note on Distributed Computing](https://scholar.harvard.edu/files/waldo/files/waldo-94.pdf) does cover these in depth. I'll think more about which ones we want to mention.

File: website/blog/corba.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+  method call act like local method call will not end well.
+- Take distributed shared memory, for example, where a binary is built and run
+  like a regular binary, but it's low-level loads and stores are executed
```
[the remark this answers is not in the document]
Comment (mwhittaker) [reply website/blog/corba.md]: Oops, thanks.

File: website/blog/corba.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+  against remote memory. DSM never took off because hiding the latency and
+  partial failures of the remote memory access is impossible.
```
[the remark this answers is not in the document]
Comment (mwhittaker) [reply website/blog/corba.md]: Good point. Softened the wording to "extremely challenging".

File: website/blog/corba.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+- NFS hard mounts are another example taken from [the paper][a_note_on_dc]. If
```
[the remark this answers is not in the document]
Comment (mwhittaker) [reply website/blog/corba.md]: Done.

File: website/blog/corba.md [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+  an operation on a hard mounted NFS volume fails, the NFS client tries to mask
+  the failure by retrying the operation indefinitely until it succeeds. This
+  means that if any NFS server fails, all clients freeze.
+- While Service Weaver remote method calls look like local method calls, they do
+  not act like them. All the realities of executing methods over the network are
+  exposed to the developer.
```
[the remark this answers is not in the document]
Comment (mwhittaker) [reply website/blog/corba.md]: Perfect. Thank you!

File: website/blog/corba.md [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+- For example, every Service Weaver method is required to return an error. When
+  a method call fails to execute properly, a `RemoteCallError` is returned,
+  allowing a developer to detect and react to failures.
+- Every Service Weaver method is also required to take a `context.Context` as
+  its first argument. Developers can set deadlines on these contexts to detect
+  and react to method calls that are experiencing high latency.
+- We are also working on extending our unit test framework to automatically
+  inject latency and failures to further ensure that Service Weaver apps work
+  well even when the network isn't.
```
[the remark this answers is not in the document]
Comment (mwhittaker) [reply website/blog/corba.md]: Done.

File: website/blog/corba.md [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+  calls are forced to have the complexity of remote method calls". This is not
+  true, as only component methods are remote. You are not forced to make every
+  method call a remote method call. However, it is a valid point that we
+  shouldn't tell people to split as much as they want without thought. There
+  should still be thought in splitting into components.
+- TODO(mwhittaker): Another counter argument is "well, what's the point of
+  unifying things then?" Service Weaver still has a number of benefits, even
+  though we don't magically solve issues of latency and failures. We could
+  explain this counterargument and rebut it, but maybe it's better to stay terse
+  and not mention it.
```
[the remark this answers is not in the document]
Comment (mwhittaker) [reply website/blog/corba.md]: :)

File: website/blog/corba.md [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+
+## Why Did CORBA Fail?
+
+- I can't speak authoritatively on why CORBA failed. It was invented before I
```
[the remark this answers is not in the document]
Comment (mwhittaker) [reply website/blog/corba.md]: Dropped!

File: website/blog/corba.md [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+  was born and fell out of popularity when I was in elementary school.
+- However, I can defer to those with expertise. [Michi Henning][michi],
+  co-author of the book [*Advanced CORBA Programming with C++*][corba_book]
+  wrote an article for ACM Queue in 2006 documenting why CORBA failed.
+- He explains that CORBA failed because it was plagued with technical and
+  bureaucratic issues. It was these issues that sank CORBA, not the distributed
+  object paradigm.
+- In fact, projects like [Ray][ray], [Microsoft Orelans][orleans], and
+  [Akka][akka] show that similar paradigms can be implemented with incredible
+  success.
```
[the remark this answers is not in the document]
Comment (mwhittaker) [reply website/blog/corba.md]: Done. I linked to the pathways paper on arxiv. Let me know if there's a more appropriate link. I saw Jeff wrote a blog post about it as well.

**Review by mwhittaker — COMMENTED** [review]
Thanks for the amazing feedback Sanjay! I integrated everything into the sketch. I'll start working on converting stuff to full prose.

**Comment by mwhittaker** [comment]
Just pushed a full draft of the blog!

File: website/blog/corba.md [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+components interact by calling these methods. When a Service Weaver application
+is deployed, some components are co-located in the same process, and others are
+run in separate processes, potentially on separate machines. As a result, some
+method calls are performed locally, and some are performed remotely.
+
+This paradigm might sound familiar to you. It's similar to the distributed
+object model used by CORBA, DCOM, Java RMI, and countless other frameworks that
+have come and gone. It is [widely believed][why_corba_failed] that these
+frameworks failed because they attempted to make remote method calls **look**
+and **act** like local method calls. This flies in the face of the [fallacies of
```
Comment (spetrovic77) [inline website/blog/corba.md:20]: The two sentences starting with `"This flies in the face..."` are making a big cognitive jump from the "look and act"  sentence. i.e., it's not at all obvious what "this" is and why it "flies in the face".

Can you just say that:

`Related research [fallacies] [a_note_on_dc] seemingly supports this theory.`

, or something to that effect?

File: website/blog/corba.md [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+
+No.
+
+The hypothesis that CORBA failed because method calls looked like local calls
+but behaved like remote calls is incorrect. CORBA actually failed for different
+reasons, which we explain below. Many successful systems have made remote calls
+look local (e.g., [TensorFlow][tensorflow], [Microsoft Orleans][orleans]). We
+hope for Service Weaver to join this list of successful systems as it gets
+adopted.
+
```
Comment (spetrovic77) [inline website/blog/corba.md:34]: Here would be a nice place to say what you are going to do in this blog post.

File: website/blog/corba.md [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+  method call serializable, which is enforced by the [Service Weaver code
+  generator][generator].
+
+Yes, these are a burden on the programmer, but that is precisely the point of
+[*A Note on Distributed Computing*][a_note_on_dc]; it is a mistake to try to
+isolate the programmer from the fundamental issues of remote calls. We are also
+working on extending our [unit test framework][testing] to automatically inject
+latency and failures to further ensure that Service Weaver applications are
+resilient to the failures and delays they will encounter in a real deployment.
+
```
Comment (spetrovic77) [inline website/blog/corba.md:154]: Might be a good place to put in the "networked disk" addendum here, espousing the speed and the resilience of modern cloud networks. WDYT?

**Review by spetrovic77 — APPROVED** [review]

**Review by ghemawat — APPROVED** [review]

**Pull request merged**